Adds Parameter use_enhanced and model to GoogleCloudSpeech (Fix #734)

HideyoshiNakazone opened the pull request:
Adds the parameters use_enhanced and model to the recognize_google_cloud method for more customizable options for the user and better results in specific cases

HideyoshiNakazone commented:
Hello @Uberi i was wondering if it's possible for someone to review my merge request.

Thank you very much, 
Vitor Hideyoshi.

Uberi commented:
Hi @HideyoshiNakazone!

Looks good overall, but would it be possible to document these parameters in the docs for that function? If so, happy to merge this!

HideyoshiNakazone commented:
@Uberi, thanks a lot! I added the parameters to the Docstring of the method `Recognizer.recognize_google_cloud` and added them to the library reference file. 
If there is any other places you'd like me to add documentation i'll be happy to :)

ftnext commented on reference/library-reference.rst:
```diff
@@ -238,6 +238,10 @@ The recognition language is determined by ``language``, which is a BCP-47 langua
 
 If ``preferred_phrases`` is an iterable of phrase strings, those given phrases will be more likely to be recognized over similar-sounding alternatives. This is useful for things like keyword/command recognition or adding new phrases that aren't in Google's vocabulary. Note that the API imposes certain `restrictions on the list of phrase strings <https://cloud.google.com/speech/limits#content>`__.
 
+The ``use_enhanced`` is a boolean option that sets a flag with the same name on the Google Cloud Speech API, it will make the API uses the enhanced version of the model. More information can be found in the `Google Cloud Speech API documentation <https://cloud.google.com/speech-to-text/docs/enhanced-models>` __.
```
@HideyoshiNakazone Thanks! Would you like to remove space?

```diff
-<https://cloud.google.com/speech-to-text/docs/enhanced-models>` __
+<https://cloud.google.com/speech-to-text/docs/enhanced-models>`__
```

ftnext commented:
@HideyoshiNakazone Thank you very much for this pull request! I'm very sorry to respond too late.
@Uberi Thanks your comment!

In my opinion, it seems to be better to introduce keyword arguments (a.k.a. `**kwargs`)
https://docs.python.org/3/tutorial/controlflow.html#keyword-arguments

Certainly, adding `use_enhanced` and `model` as arguments would implement this feature.
However, if there are additional arguments to be added in the future, there is a concern that they could be added again (not easy to extend).

I think it would be preferable for Cloud Speech API-specific arguments to be specified as variant keyword arguments.

```python
def recognize_google_cloud(self, audio_data, credentials_json=None, language="en-US", preferred_phrases=None, show_all=False, **api_params):
    """
    If ``preferred_phrases`` is an iterable of phrase strings, ...

    api_params: Cloud Speech API-specific parameters as dict (optional)

        The ``use_enhanced`` is a boolean option ...

        Furthermore, you can use the option ``model`` to set your desired model,

    Returns the most likely transcription if ``show_all`` is False (the default).
    """

    config = {
        'encoding': speech.RecognitionConfig.AudioEncoding.FLAC,
        'sample_rate_hertz': audio_data.sample_rate,
        'language_code': language,
        **api_params,
    }
```

(It seems that `preferred_phrases` might be included in `api_params` too, but this is another issue)

HideyoshiNakazone commented:
Hi @ftnext and @Uberi ,

I know it has been a while, but could you please take another look at my pull request? I believe it now meets the requirements discussed in this thread, and it would be great if we could proceed with merging it.

Thanks for your time!

ftnext commented on speech_recognition/__init__.py:
```diff
@@ -713,6 +724,13 @@ def recognize_google_cloud(self, audio_data, credentials_json=None, language="en
         assert isinstance(language, str), "``language`` must be a string"
         assert preferred_phrases is None or all(isinstance(preferred_phrases, (type(""), type(u""))) for preferred_phrases in preferred_phrases), "``preferred_phrases`` must be a list of strings"
 
+        # Implementation of assertions of common api_params
```
Could you remove assertions?

ref: https://realpython.com/python-assert-statement/#when-not-to-use-assertions
>In general, you shouldn’t use assertions for data processing or data validation

This repository has repeatedly made the mistake of using assertions for data validation, and I don't want to add new ones as a maintainer

HideyoshiNakazone replied:
That's fair, i removed the antipattern. I hope the code is now acceptable :)

ftnext reviewed (changes requested):
Thanks!
May I make a simple correction?

ftnext reviewed (approved):
Thanks!
